src/capacity.rs

```diff
@@ -0,0 +1,164 @@
+    pub fn timestamp(&self) -> f64 {
+        self.timestamp
+    }
+
+    /// An average of collected load values over the given update interval
+    pub fn average_load(&self) -> f64 {
+        self.average_load
+    }
+
+    /// A maximum value from collected load values over the given update interval.
+    pub fn peak_load(&self) -> f64 {
+        self.peak_load
+    }
+
+    /// A ratio between the number of buffer underruns and the total number of system-level audio callbacks over the given update interval.
+    pub fn underrun_ratio(&self) -> f64 {
+        self.underrun_ratio
+    }
+}
+
+type EventHandler = Box<dyn FnMut(AudioRenderCapacityEvent) + Send + 'static>;
+
+/// Provider for rendering performance metrics
+///
+/// A load value is computed for each system-level audio callback, by dividing its execution
+/// duration by the system-level audio callback buffer size divided by the sample rate.
+///
+/// Ideally the load value is below 1.0, meaning that it took less time to render the audio than it
+/// took to play it out. An audio buffer underrun happens when this load value is greater than 1.0: the
+/// system could not render audio fast enough for real-time.
+pub struct AudioRenderCapacity {
+    sender: Sender<LoadValueData>,
+    receiver: Option<Receiver<LoadValueData>>,
+    callback: Arc<Mutex<Option<EventHandler>>>,
+}
+
+impl AudioRenderCapacity {
+    pub(crate) fn new(sender: Sender<LoadValueData>, receiver: Receiver<LoadValueData>) -> Self {
+        let callback = Arc::new(Mutex::new(None));
+
+        Self {
+            sender,
+            receiver: Some(receiver),
+            callback,
+        }
+    }
+
+    /// Start metric collection and analysis
+    #[allow(clippy::missing_panics_doc)]
+    pub fn start(&mut self, options: AudioRenderCapacityOptions) {
+        let receiver = match self.receiver.take() {
+            None => return,
+            Some(receiver) => receiver,
+        };
+
+        let callback = self.callback.clone();
+
+        let mut timestamp: f64 = 0.;
+        let mut load_sum: f64 = 0.;
+        let mut counter = 0;
+        let mut peak_load: f64 = 0.;
+        let mut underrun_sum = 0;
+
+        let mut next_checkpoint = options.update_interval;
+        std::thread::spawn(move || {
+            for item in receiver {
+                let LoadValueData {
+                    render_timestamp,
+                    load_value,
+                } = item;
+
+                // check for signal to stop
+                if render_timestamp.is_nan() && load_value.is_nan() {
+                    return; // stop thread
+                };
+
+                counter += 1;
+                load_sum += load_value;
+                peak_load = peak_load.max(load_value);
+                if load_value > 1. {
+                    underrun_sum += 1;
+                }
+
+                if render_timestamp >= next_checkpoint {
+                    let event = AudioRenderCapacityEvent::new(
```

Comment:
I understand there is probably no need to limit the precision here (problems of fingerprinting are not the same), but maybe we could just add a comment referring to (https://webaudio.github.io/web-audio-api/#dom-audiorendercapacityevent-averageload)?

Reply:
done